### PR TITLE
feature: spinlock benchmark

### PR DESCRIPTION
**Pull request by TaiJuWu**
## Summary
spinlock benchmark
## Impact
None
## Testing
local test

### CODE REVIEW

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+void spinlock_bench_main(void)
+{
+  int status;
+  g_result = 0;
+
+  clock_t start = perf_gettime();
+  for (int i = 0; i < THREAD_NUM; ++i)
+    {
+      status = pthread_create(&g_thread[i], NULL,
+                            (void *)thread_spinlock, &g_thread[i]);
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:71]: remove the cast and align

File: benchmarks/spinlock_bench/Makefile [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+include $(APPDIR)/Make.defs
+
+# spinlock_bench application
+
+############################################################################
+# Targets
+############################################################################
+
+clean::
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/Makefile:29]: why need

File: benchmarks/spinlock_bench/Makefile [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	$(call DELFILE, $(OBJS))
+
+# Download and unpack tarball if no git repo found
+
+distclean::
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/Makefile:34]: ditto

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+#define TEST_NUM CONFIG_SPINLOCK_MULTITHREAD
+#define THREAD_NUM CONFIG_SPINLOCK_MULTITHREAD
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static spinlock_t lock = SP_UNLOCKED;
+
+static pthread_t g_thread[THREAD_NUM];
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:44]: change to local

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+static int g_result = 0;
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:46]: let's pass g_result and lock through thread's parameter

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      g_result++;
+      spin_unlock(&lock);
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void spinlock_bench_main(void)
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:62]:
```suggestion
void main(void)
```

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+static pthread_t g_thread[THREAD_NUM];
+
+static int g_result = 0;
+
+static FAR void thread_spinlock(FAR void *parameter)
+{
+  for (int i = 0; i < TEST_NUM; i++)
+    {
+      spin_lock(&lock);
```
Comment (pkarashchenko) [inline benchmarks/spinlock_bench/spinlock_bench.c:52]: Maybe better use `pthread_spin_lock`?

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void spinlock_bench_main(void)
+{
+  int status;
```
Comment (pkarashchenko) [inline benchmarks/spinlock_bench/spinlock_bench.c:64]:
```suggestion
  int status;
  int i;
```

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  g_result = 0;
+
+  clock_t start = perf_gettime();
+  for (int i = 0; i < THREAD_NUM; ++i)
```
Comment (pkarashchenko) [inline benchmarks/spinlock_bench/spinlock_bench.c:68]:
```suggestion
  for (i = 0; i < THREAD_NUM; ++i)
```

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    {
+      status = pthread_create(&g_thread[i], NULL,
+                            (void *)thread_spinlock, &g_thread[i]);
+      if (status != 0)
+        {
+          printf("spinlock_test: ERROR pthread_create failed, status=%d\n",
+                  status);
```
Comment (pkarashchenko) [inline benchmarks/spinlock_bench/spinlock_bench.c:75]:
```suggestion
          printf("spinlock_test: ERROR pthread_create failed, status=%d\n",
                 status);
```

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 7 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: The implement of pthread_spin_lock is different with kernel spinlock.
My initial idea is testing kernel spinlock.

Comment (pkarashchenko) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: Ok.

Comment (pkarashchenko) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: So this test will be enabled for FLAT build only?

File: benchmarks/spinlock_bench/Makefile (hunk 2 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/Makefile:29]: After checking, it can be deleted.

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 7 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: > So this test will be enabled for FLAT build only?

Yes.

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 5 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c]: Done.

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+#define TEST_NUM CONFIG_SPINLOCK_MULTITHREAD
+#define THREAD_NUM CONFIG_SPINLOCK_MULTITHREAD
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct thread_parmeter_s
+{
+  int *g_result;
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:44]:
```suggestion
  int *result;
```

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define TEST_NUM CONFIG_SPINLOCK_MULTITHREAD
+#define THREAD_NUM CONFIG_SPINLOCK_MULTITHREAD
+
+/****************************************************************************
+ * Private Data
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:39]: Type

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+ ****************************************************************************/
+
+struct thread_parmeter_s
+{
+  int *g_result;
+  spinlock_t *lock;
+};
+
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:47]: add private function banner

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+static FAR void *thread_spinlock(FAR void *parameter)
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:48]: let's use FAR consistent across the code base.

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void main(void)
+{
+  spinlock_t lock = SP_UNLOCKED;
+  int g_result = 0;
+  pthread_t g_thread[THREAD_NUM];
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:71]: removr ALL g_ prefix

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 14 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: Im not sure every type is needed to add FAR or not.
If there is any mistake, please tell me.

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 13 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c:47]: Is this called banner?
```
/****************************************************************************
 * Private Functions
 ****************************************************************************/
 ```

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 15 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c]: Done.

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 14 above)
Comment (xiaoxiang781216) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: pointer in common code need add FAR, except the arch/chip/board specific code could remove FAR if they don't distinguish near and far.

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static FAR void *thread_spinlock(FAR void *parameter)
+{
+  FAR int *result = ((FAR struct thread_parmeter_s *)parameter)->result;
+  spinlock_t *lock = ((FAR struct thread_parmeter_s *)parameter)->lock;
+
+  for (FAR int i = 0; i < TEST_NUM; i++)
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:57]: remove FAR from non pointer

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+  return NULL;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void main(void)
+{
+  FAR spinlock_t lock = SP_UNLOCKED;
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:73]: remove ALL FAR from all non pointer.

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+          ASSERT(false);
+        }
+    }
+
+  for (i = 0; i < THREAD_NUM; ++i)
+    {
+      pthread_join(thread[i], NULL);
+    }
+
+  FAR clock_t end = perf_gettime();
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:102]: Move `clock_t end` to the begin

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 14 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c:52]: > pointer in common code need add FAR, except the arch/chip/board specific code could remove FAR if they don't distinguish near and far.

Got it. Thanks your answer.

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 18 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c]: Done.

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+  FAR spinlock_t *lock;
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static FAR void *thread_spinlock(FAR void *parameter)
+{
+  FAR int *result = ((struct thread_parmeter_s *)parameter)->result;
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:54]:
```suggestion
  FAR int *result = ((FAR struct thread_parmeter_s *)parameter)->result;
```

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+  FAR spinlock_t *lock = ((struct thread_parmeter_s *)parameter)->lock;
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:55]:
```suggestion
  FAR spinlock_t *lock = ((FAR struct thread_parmeter_s *)parameter)->lock;
```

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+void main(void)
+{
+  spinlock_t lock = SP_UNLOCKED;
+  int result = 0;
+  pthread_t thread[THREAD_NUM];
+  struct thread_parmeter_s para;
+  clock_t start, end;
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:77]: split to.two line

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+  int i;
+
+  para.result = &result;
+  para.lock = &lock;
+
+  start = perf_gettime();
+  for (i = 0; i < THREAD_NUM; ++i)
+    {
+      status = pthread_create(&thread[i], NULL,
+                            thread_spinlock, &para);
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:89]:
```suggestion
                              thread_spinlock, &para);
```

File: benchmarks/spinlock_bench/spinlock_bench.c (hunk 21 above)
Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c]: Done. Please review.
Thank you a lot!!!

File: benchmarks/spinlock_bench/spinlock_bench.c [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static FAR void *thread_spinlock(FAR void *parameter)
+{
+  FAR int *result = ((FAR struct thread_parmeter_s *)parameter)->result;
+  FAR spinlock_t *lock = ((FAR struct thread_parmeter_s *)parameter)->lock;
+
+  for (int i = 0; i < TEST_NUM; i++)
```
Comment (xiaoxiang781216) [inline benchmarks/spinlock_bench/spinlock_bench.c:57]: `int i` move out of for loop

Comment (TaiJuWu) [reply benchmarks/spinlock_bench/spinlock_bench.c]: Done.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**